### PR TITLE
Top level generation script for PR, MC, TC, TB

### LINKED ISSUE #177: Scripts to generate top functions
There is currently a Python script that can generate the supported top functions of the TC from the wiring configuration:
https://github.com/cms-L1TK/firmware-hls/blob/master/emData/generate_TC.py
This script also generates all the bit masks and other constants that are unique to each instance of the TC.

All of the modules should have similar scripts written for them that generate top functions from a given wiring configuration, at least for the instances that will appear in the summer chain (L1L2 seeds with matches only in the barrel).

### PR DESCRIPTION
**Pull request by bryates**
Resolves #177 for MC, other modules can push to this PR.

### CODE REVIEW

File: emData/generate_MC.py
Comment (aehart) [inline emData/generate_MC.py:106]: @bryates This is nitpicky, but could you use the same suffix style as is used in the wiring, e.g. something like `MatchCalculator_L3PHIC` instead of `MatchCalculator_L3C`?

**Comment by aehart** [comment]
@bryates Thanks for responding so quickly.

Could you please also replace the files in the repo with those generated by generate_MC.py, run on the current wiring (i.e., that in emData/LUTs/wires.dat after running emData/download.sh)? I think all the modules should do this in order to verify that the scripts produce correct output.

Also, are you planning to expand the script to generate other MC top functions? For the reduced config, we'll need MCs for L3, L4, L5, and L6. And the current reduced config uses MC_L*PHIB, so we'll need support for that phi sector at least, though generating top functions for all phi sectors would be ideal.

File: TrackletAlgorithm/MatchCalculator.h
Comment (bryates) [inline TrackletAlgorithm/MatchCalculator.h:245]: I added these lines to make the templating easier, but I'm not sure if this is the right file to put them in. Should we put things like this in Constants.h, or should the generation scripts be able to calculate the template parameters?

File: emData/generate_MC.py
Comment (bryates) [reply emData/generate_MC.py]: Good catch!

**Comment by bryates** [comment]
> @bryates Thanks for responding so quickly.
> 
> Could you please also replace the files in the repo with those generated by generate_MC.py, run on the current wiring (i.e., that in emData/LUTs/wires.dat after running emData/download.sh)? I think all the modules should do this in order to verify that the scripts produce correct output.
> 
> Also, are you planning to expand the script to generate other MC top functions? For the reduced config, we'll need MCs for L3, L4, L5, and L6. And the current reduced config uses MC_L*PHIB, so we'll need support for that phi sector at least, though generating top functions for all phi sectors would be ideal.

I've replaced the default files with the generated ones. Should we modify the script to directly place them in the correct directory, instead of in `emData`?
I'll start playing with the other layers and phi sector B.

**Comment by aehart** [comment]
> I've replaced the default files with the generated ones. Should we modify the script to directly place them in the correct directory, instead of in `emData`?

Yeah, I think it makes sense to do that. Good suggestion.

File: TrackletAlgorithm/MatchCalculator.h
Comment (aehart) [reply TrackletAlgorithm/MatchCalculator.h:245]: As we move to automatically generating the top functions, I think the simplest thing is to have the scripts calculate these parameters directly. So, e.g., instead of the scripts generating a call to `ASRegion<TF::L3>()`, they would just write `BARRELPS` where appropriate. I'll need to update generate_TC.py to be consistent with this approach, but I think it's the way we should go.

Comment (bryates) [reply TrackletAlgorithm/MatchCalculator.h]: Done.

**Comment by bryates** [comment]
> > @bryates Thanks for responding so quickly.
> > Could you please also replace the files in the repo with those generated by generate_MC.py, run on the current wiring (i.e., that in emData/LUTs/wires.dat after running emData/download.sh)? I think all the modules should do this in order to verify that the scripts produce correct output.
> > Also, are you planning to expand the script to generate other MC top functions? For the reduced config, we'll need MCs for L3, L4, L5, and L6. And the current reduced config uses MC_L*PHIB, so we'll need support for that phi sector at least, though generating top functions for all phi sectors would be ideal.
> 
> I've replaced the default files with the generated ones. Should we modify the script to directly place them in the correct directory, instead of in `emData`?
> I'll start playing with the other layers and phi sector B.

I've tested the updates with `L3PHIB` and `L5PHIB`, and both work correctly.

**Comment by pwittich** [comment]
This code is not safe for python 3. The fixes are trivial.

1. add `from __future__ import absolute_import` to each file
2. add `from __future__ import print_function` and add parens around each print call, and resolve any remaining complaints.

```
(py2) ➜  emData git:(generate_top) pylint --py3k generate_*.py
No config file found, using default configuration
************* Module generate_MC
W:  7, 0: import missing `from __future__ import absolute_import` (no-absolute-import)
E: 27, 4: print statement used (print-statement)
************* Module generate_TC
W:  8, 0: import missing `from __future__ import absolute_import` (no-absolute-import)
E: 64, 4: print statement used (print-statement)

------------------------------------------------------------------
Your code has been rated at 9.45/10 (previous run: 9.60/10, -0.15)
```

File: emData/generate_MC.py
Comment (aehart) [inline emData/generate_MC.py:43]: @bryates Is there any reason we can't generate top functions for all MCs, or at least all barrel MCs? We don't need to test them all right now, but it would be nice if they were all defined.

File: emData/generate_MC.py
Comment (aehart) [inline emData/generate_MC.py:45]: @bryates Please change `tcName` → `mcName`.

File: emData/generate_MC.py
Comment (aehart) [inline emData/generate_MC.py:63]: @bryates Please change `TrackAlgorithm` → `TrackletAlgorithm`

File: emData/generate_MC.py
Comment (aehart) [inline emData/generate_MC.py:78]: @bryates Please change `TrackAlgorithm` → `TrackletAlgorithm`

File: emData/generate_MC.py
Comment (aehart) [inline emData/generate_MC.py:97]: @bryates Please change `tcName` → `mcName`.

File: emData/generate_PR.py
Comment (aehart) [inline emData/generate_PR.py:1]: @djcranshaw Please add the following imports to ensure Python3 compatibility:
```python
from __future__ import absolute_import, print_function
```
They have to be included before anything else.

File: emData/generate_PR.py
Comment (aehart) [inline emData/generate_PR.py:5]: @djcranshaw Is there any reason we can't expand this list to include all PRs, or at least all barrel PRs? We don't need to test them all right now, but it would be nice if they were all defined.

File: emData/generate_PR.py
Comment (aehart) [inline emData/generate_PR.py:9]: @djcranshaw Close `wiresFile` after reading it.

File: emData/generate_PR.py
Comment (aehart) [inline emData/generate_PR.py:75]: @djcranshaw I guess this function will return nonsense if the module is not in the given wiring? I think either it should return an empty string when `TProjCount` and `VMProjCount` are zero, or you should generate `modules_to_test` from the wiring itself.

File: emData/generate_MC.py
Comment (bryates) [reply emData/generate_MC.py:45]: Done. I'll push the change soon.

File: emData/generate_MC.py
Comment (bryates) [reply emData/generate_MC.py:78]: Done. I'll push the change soon.

File: emData/generate_MC.py
Comment (bryates) [reply emData/generate_MC.py:97]: Done. I'll push the change soon.

File: emData/generate_MC.py
Comment (bryates) [reply emData/generate_MC.py]: Done. I just have disks disabled now.
https://github.com/cms-L1TK/firmware-hls/blob/a4edeb22db02d914e2987074766de0f12aedbd19/emData/generate_MC.py#L40

File: emData/generate_MC.py
Comment (bryates) [reply emData/generate_MC.py]: Done.
https://github.com/cms-L1TK/firmware-hls/blob/a4edeb22db02d914e2987074766de0f12aedbd19/emData/generate_MC.py#L59-L60

File: emData/generate_MC.py
Comment (bryates) [inline emData/generate_MC.py:98]: @aehart is this ok? I basically made this PR hang on #179. If this one is merged first, I can just push this change to the #179, otherwise I can push it here.

**Comment by aehart** [comment]
FYI, I just rebased and removed the commits related to the MatchProcessor, so as to expedite the review for modules in the summer chain.

@bryates I moved the MP commits to the mp_generate_top branch. Please wait until this PR is merged to open a PR for that branch.

**Comment by meisonlikesicecream** [comment]
BTW, do we want to add something about the top level scripts in the README?

File: emData/generate_PR.py
Comment (djcranshaw) [reply emData/generate_PR.py:77]: @aehart Yes, you're right that this function would return nonsense if the module isn't in the given wiring. This is related to the above comment on modules_to_build as well - I had originally envisioned that variable to just hold the names of modules that we know will build top-levels that work, but maybe the right approach is just to build all modules out of the box, and only add the modules that we want to test into the tcl script. And if we do that, there's no need to list the modules individually in their own variable, they can just be built from the wiring file. This approach would mean that we build some top-levels that don't work, but if we're not testing/using them yet anyways maybe that's fine. Also all the modules needed for the summer chain do work.

File: emData/generate_PR.py
Comment (djcranshaw) [reply emData/generate_PR.py:4]: Thanks, done.

File: emData/generate_PR.py
Comment (djcranshaw) [reply emData/generate_PR.py:17]: Thanks, done.

**Review by aehart — APPROVED** [review]
Looks good with the most recent changes. Approved!

**Pull request merged**
